Adin1110 add support

tachicialex opened the pull request:
The ADIN1110 is a low power single port 10BASE-T1L MAC-
PHY designed for industrial Ethernet applications. It integrates
an Ethernet PHY core with a MAC and all the associated analog
circuitry, input and output clock buffering.

The device has a 4-wire SPI interface for communication
between the MAC and host processor.

68a8ec0c5d95 - dt-bindings: net: adin1110: Add docs:
DT-bindings for ADIN1110 MAC-PHY

0dbe577e4bf4 - net: adi: adin1110: Add initial support:
This adds the necessary net_device_ops and basic ethtool support.

6706a27fa3a7 - net: phy: adin1100: Add ADIN1110 PHY_ID:
PHY ID of the ADIN1100 phy that is present in the ADIN1110 MAC-PHY chip differs from the one in the standalone version

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin1110.yaml:
```diff
@@ -0,0 +1,98 @@
+
+  reg:
+    description: |
+      SPI CS number
```
nit: typically no need for a description... Also don't we need `maxItems: 1`?

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin1110.yaml:
```diff
@@ -0,0 +1,98 @@
+  adi,spi-crc:
+    description: |
+      Enable CRC8 checks on SPI read/writes.
+    type: boolean
```
Arguably this could always be on... not sure if this really fits in a devicetree property. At least on IIO, I remember some complains about this stuff. But let's see what they have to say in netdev (maybe they are fine with it)

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin1110.yaml:
```diff
@@ -0,0 +1,98 @@
+
+  interrupts:
+    description: |
+      Used to signal on falling edge a status change in adin1110.
```
I would also remove the description here... Also, shouldn't we have a maxItems for this?

nunojsa commented on drivers/net/ethernet/adi/Kconfig:
```diff
@@ -0,0 +1,28 @@
+	tristate "Analog Devices ADIN1110 MAC-PHY"
+	depends on SPI
+	select CRC8
+	---help---
```
Just use `help`. This syntax is deprecated and will start trigger a build error as we move to 5.10

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	netdev->netdev_ops = &adin1110_netdev_ops;
+	netdev->ethtool_ops = &adin1110_ethtool_ops;
+
+	ret = register_netdev(netdev);
```
`devm_register_netdev()`?

tachicialex replied:
This one is added in 5.8.

nunojsa replied:
I see. Oh well, it's a change you can do upstream and we will have this in 5.10

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	priv->phydev = phy_find_first(priv->mii_bus);
+	if (!priv->phydev) {
+		adin1110_unregister_mdiobus(priv);
+
```
nit: I would not use an extra blank line here and in all other places like this...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	ret = adin1110_check_spi(priv);
+	if (ret < 0) {
+		netdev_err(netdev, "SPI read failed: %d\n", ret);
+		mutex_unlock(&priv->lock);
```
uh? Am I missing something? Why the `mutex_unlock()`?

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	mii_bus->phy_mask = ~((u32)BIT(0));
+	snprintf(mii_bus->id, MII_BUS_ID_SIZE, "%s", dev_name(dev));
+
+	ret = mdiobus_register(mii_bus);
```
`devm_mdiobus_register()`?

tachicialex replied:
devm_mdiobus_register is added in 5.8

nunojsa replied:
I see. The same as for the other API. I would do this change on the version you'll send upstream and we will also get these as soon as we move to 5.10

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	struct mii_bus *mii_bus;
+	int ret;
+
+	mii_bus = mdiobus_alloc();
```
`devm_mdiobus_alloc()`?

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	adin1110_unregister_mdiobus(priv);
+
+	return 0;
+}
```
It seems that we can use devm variants and the remove callback can be skipped...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	struct adin1110_priv *priv = netdev_priv(dev);
+	netdev_tx_t netdev_ret = NETDEV_TX_OK;
+
+	spin_lock(&priv->state_lock);
```
curious: Why the spinlock? Is this running from IRQ context?

tachicialex replied:
No sleeping allowed in ndo_start_xmit and ndo_set_rx_mode. Softirqs are disabled here. Tried it and it crashed badly.

nunojsa replied:
Hmm I see...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+	adin1110_read_reg(priv, ADIN1110_STATUS1, &status1);
+
+	if (priv->append_crc && (status1 & ADIN1110_SPI_ERR))
+		netdev_err(priv->netdev, "SPI CRC error on write.\n");
```
Shouldn't we do something on error? Otherwise, I would use `netdev_warn()`

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+
+		usleep_range(10000, 15000);
+		timeout -= 10;
+	}
```
Maybe `readx_poll_timeout()` can be used here...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,1017 @@
+
+		usleep_range(10000, 15000);
+		timeout -= 10;
+	}
```
Maybe also here...

nunojsa reviewed (commented):
As we talked about, we are going first try to upstream this. Anyways, FWIW, just some minor feedback from my side

tachicialex commented:
Changelog v1 -> v2:
- removed from yaml the reg description, added maxItems: 1
- maxItems: 1 added to interrupts
- help instead of --help--
- devm_mdiobus_alloc instead of mdiobus_alloc
- netdev_warn on SPI CRC error
- removed wrong mutex_unlock from probe()
- added readx_poll_timeout instead of custom timeout stuff
- added to Kconfig.adi

mhennerich commented:
closed? Why?

tachicialex commented:
> closed? Why?

Will attempt to send it upstream first, after I finish upstreaming adin1100.
